util/src/main/java/org/wildfly/security/util/PasswordUtil.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2021 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.wildfly.security.util;
+
+import org.wildfly.common.iteration.ByteIterator;
+
+import java.security.SecureRandom;
+
+public class PasswordUtil {
+
+    private static SecureRandom random = new SecureRandom();
+
+    /**
+     * Generate a random salt as byte array.
+     *
+     * @param saltSize the size of the salt
+     * @return a byte array representing the random salt
+     */
+    public static byte[] generateSecureRandomSalt(int saltSize) {
+        byte[] salt = new byte[saltSize];
+        random.nextBytes(salt);
+        return salt;
+    }
+
+    /**
+     * Generate a random salt as String of given byte size.
+     *
+     * @param saltSize the size of the salt
+     * @return a string representing the random salt
+     */
+    public static String generateSecureRandomSaltString(int saltSize) {
+        byte[] salt = generateSecureRandomSalt(saltSize);
+        String saltString = ByteIterator.ofBytes(salt).base64Encode().limitedTo(saltSize).drainToString();
```

Comment:
Thanks for the updates!

Here, I think you could use `return new String(salt, StandardCharsets.UTF_8)` instead.

Reply:
Hi @fjuma, at the first I have tried a similar approach, however straight conversion, with UTF-8 encoding, of randomly generated bytes frequently resulted in unreadable strings full of special characters and whitespaces. So I have tried Base64 encoding, as it effectively limits the character set. To be honest, I am not sure if readable string as the output of the mask command is in fact needed.

Reply:
Hi @lvydra, ah, I see what you mean. The output of the mask command will get used when configuring a `masked-password` (e.g., see https://developer.jboss.org/people/aabdelsa/blog/2019/09/05/support-for-masked-passwords-in-the-client-xml-configuration for an example). So the salt should be a readable string that can be copy/pasted when configuring the `masked-password`.

What if we tried something like this:

```
private static final SecureRandom RANDOM = new SecureRandom();
private static final String CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

static String generateRandomString(int length) {
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < length; i++) {
            int index = (int) (RANDOM.nextDouble() * CHARS.length());
            builder.append(CHARS.substring(index, index + 1));
        }
        return builder.toString();
}
```

Reply:
Hi @fjuma, thanks, updated.